pkg/tagger/remote/tagger.go

```diff
@@ -268,6 +268,13 @@ func (t *Tagger) run() {
 }
 
 func (t *Tagger) processResponse(response *pb.StreamTagsResponse) error {
+	// returning early when there are no events prevents a keep-alive sent
+	// from the core agent from wiping the store clean in case the remote
+	// tagger was previously in an unready (but filled) state.
+	if len(response.Events) == 0 {
+		return nil
+	}
+
```

Comment:
So, the assumption is that a StreamTagsResponse either contains _all_ tags or no tags, with the no-tags option being a keepalive?

Reply:
Slightly more complicated. The first response on a stream is a bundle of synthetic events that set all tags. Any event that comes afterwards sets and unsets tags as they come and go from the store in the core agent. An empty response is either a keep-alive, or an initial bundle with nothing, because there are no tags, there's no other way for an empty event to be generated.